sale_coupon_advanced/models/sale_coupon.py

```diff
@@ -0,0 +1,38 @@
+import secrets
+import string
+
+from odoo import _, api, fields, models
+from odoo.exceptions import UserError
+
+
+class SaleCoupon(models.Model):
+    """Extend to implement custom code generation rule."""
+
+    _inherit = "sale.coupon"
+
+    # Must overwrite here for override method to take effect.
+    code = fields.Char(default=lambda s: s._generate_code())
+
+    def _validate_custom_code_rule(self, custom_code_rule):
```

Comment:
Why not overwriting original `_check_coupon_code` here to extends the original sale_coupon module ?

Reply:
Because here I am checking code generation rule, not code/coupon itself. And if I would override that, I would need to either introduce new argument in method or pass it via context.